File: tests_deps/src/test/java/com/github/klee0kai/stone/test/cachecontrol/gc/WindowPartialGcTests.java
package com.github.klee0kai.stone.test.cachecontrol.gc;

import com.github.klee0kai.stone.Stone;
import com.github.klee0kai.stone.closed.types.ListUtils;
import com.github.klee0kai.test.car.di.cachecontrol.gc.CarGcComponent;
import com.github.klee0kai.test.car.model.Window;
import org.junit.jupiter.api.Test;

import java.lang.ref.WeakReference;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

public class WindowPartialGcTests {

    @Test
    public void createWorkCorrect() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);

        //When
        List<WeakReference<Window>> windowStrong = ListUtils.format(DI.windowsModule().windowStrong().get(), WeakReference::new);

        //Then
        assertEquals(3, nonNullCount(windowStrong));
    }

    @Test
    public void holdInListTest() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Window>> windowStrong = ListUtils.format(DI.windowsModule().windowStrong().get(), WeakReference::new);
        Window holder = DI.windowsModule().windowStrong().get().get(1);


        //When
        DI.gcAll();

        // Then
        assertNotNull(holder);
        assertEquals(1, nonNullCount(windowStrong));
        assertNotNull(windowStrong.get(1).get());
    }

    @Test
    public void partialRecreateList1Test() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<String> uids1 = ListUtils.format(DI.windowsModule().windowStrong().get(), it -> it.uuid);
        Window holder = DI.windowsModule().windowStrong().get().get(0);


        //When
        DI.gcAll();
        List<String> uids2 = ListUtils.format(DI.windowsModule().windowStrong().get(), it -> it.uuid);


        // Then
        assertNotNull(holder);
        assertEquals(uids1.get(0), uids2.get(0));
        assertNotEquals(uids1.get(1), uids2.get(1));
        assertNotEquals(uids1.get(2), uids2.get(2));
    }

    @Test
    public void partialRecreateList2Test() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<String> uids1 = ListUtils.format(DI.windowsModule().windowStrong().get(), it -> it.uuid);
        Window holder = DI.windowsModule().windowStrong().get().get(1);


        //When
        DI.gcAll();
        List<String> uids2 = ListUtils.format(DI.windowsModule().windowStrong().get(), it -> it.uuid);


        // Then
        assertNotNull(holder);
        assertNotEquals(uids1.get(0), uids2.get(0));
        assertEquals(uids1.get(1), uids2.get(1));
        assertNotEquals(uids1.get(2), uids2.get(2));
    }

    @Test
    public void partialRecreateList3Test() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<String> uids1 = ListUtils.format(DI.windowsModule().windowStrong().get(), it -> it.uuid);
        Window holder = DI.windowsModule().windowStrong().get().get(2);


        //When
        DI.gcAll();
        List<String> uids2 = ListUtils.format(DI.windowsModule().windowStrong().get(), it -> it.uuid);


        // Then
        assertNotNull(holder);
        assertNotEquals(uids1.get(0), uids2.get(0));
        assertNotEquals(uids1.get(1), uids2.get(1));
        assertEquals(uids1.get(2), uids2.get(2));
    }

    private int nonNullCount(List<WeakReference<Window>> list) {
        int count = 0;
        for (int i = 0; i < list.size(); i++) {
            if (list.get(i) != null && list.get(i).get() != null) count++;
        }
        return count;
    }


}

File: tests_wraps_kotlin/src/test/kotlin/com/github/klee0kai/stone/test/car/cachecontrol/gc/WindowPartialGcTests.kt
package com.github.klee0kai.stone.test.car.cachecontrol.gc

import com.github.klee0kai.stone.Stone
import com.github.klee0kai.test.car.di.cachecontrol.gc.CarGcComponent
import com.github.klee0kai.test.car.model.Window
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.lang.ref.WeakReference

class WindowPartialGcTests {
    @Test
    fun createWorkCorrect() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)

        //When
        val windowStrong = DI.windowsModule()!!.windowStrong().get()!!.map { WeakReference(it) }

        //Then
        assertEquals(3, nonNullCount(windowStrong))
    }

    @Test
    fun holdInListTest() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val windowStrong = DI.windowsModule()!!.windowStrong().get()!!.map { WeakReference(it) }
        val holder = DI.windowsModule()!!.windowStrong().get()!![1]


        //When
        DI.gcAll()

        // Then
        assertNotNull(holder)
        assertEquals(1, nonNullCount(windowStrong))
        assertNotNull(windowStrong[1]!!.get())
    }

    @Test
    fun partialRecreateList1Test() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val uids1 = DI.windowsModule()!!.windowStrong().get()!!.map { it.uuid }
        val holder = DI.windowsModule()!!.windowStrong().get()!![0]


        //When
        DI.gcAll()
        val uids2 = DI.windowsModule()!!.windowStrong().get()!!.map { it.uuid }


        // Then
        assertNotNull(holder)
        assertEquals(uids1[0], uids2[0])
        assertNotEquals(uids1[1], uids2[1])
        assertNotEquals(uids1[2], uids2[2])
    }

    @Test
    fun partialRecreateList2Test() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val uids1 = DI.windowsModule()!!.windowStrong().get()!!.map { it.uuid }
        val holder = DI.windowsModule()!!.windowStrong().get()!![1]


        //When
        DI.gcAll()
        val uids2 = DI.windowsModule()!!.windowStrong().get()!!.map { it.uuid }


        // Then
        assertNotNull(holder)
        assertNotEquals(uids1[0], uids2[0])
        assertEquals(uids1[1], uids2[1])
        assertNotEquals(uids1[2], uids2[2])
    }

    @Test
    fun partialRecreateList3Test() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val uids1 = DI.windowsModule()!!.windowStrong().get()!!.map { it.uuid }
        val holder = DI.windowsModule()!!.windowStrong().get()!![2]


        //When
        DI.gcAll()
        val uids2 = DI.windowsModule()!!.windowStrong().get()!!.map { it.uuid }


        // Then
        assertNotNull(holder)
        assertNotEquals(uids1[0], uids2[0])
        assertNotEquals(uids1[1], uids2[1])
        assertEquals(uids1[2], uids2[2])
    }

    private fun nonNullCount(list: List<WeakReference<Window?>?>): Int = list.count { it?.get() != null }
}
